azurelinuxagent/common/protocol/extensions_config_retriever.py

```diff
@@ -0,0 +1,348 @@
+
+    def _set_reason(self, reason):
+        self._reason = "{0}: FastTrack={1}, Fabric={2}".format(reason, self._ft_changed_detail, self._fabric_changed_detail)
+
+    def _get_fast_track_changed(self, artifacts_profile):
+        if artifacts_profile is None:
+            self._ft_changed_detail = "NoProfile"
+            return False
+        if not artifacts_profile.has_extensions():
+            self._ft_changed_detail = "NoExtensions"
+            return False
+
+        sequence_number = self._last_seqNo
+        if sequence_number is None:
+            sequence_number = self._get_sequence_number()
+        if sequence_number is None or sequence_number != artifacts_profile.get_sequence_number():
+            self._ft_changed_detail = "seqNoChanged"
+            return True
+
+        self._ft_changed_detail = "NoChange"
+        return False
+
+    def _get_fabric_changed(self, goal_state_incarnation):
+        if goal_state_incarnation is None:
+            self._fabric_changed_detail = "NoInc"
+            return True
+
+        incarnation = self._last_incarnation
+        if incarnation is None:
+            incarnation = self._get_incarnation()
+        if incarnation is None or str(incarnation) != str(goal_state_incarnation):
+            self._fabric_changed_detail = "IncChanged"
+            return True
+
+        self._fabric_changed_detail = "NoChange"
+        return False
+
+    def _set_fast_track(self, vm_artifacts_seq_no=None):
+        path = os.path.join(conf.get_lib_dir(), GOAL_STATE_SOURCE_FILE_NAME)
+        self._save_cache(path, GOAL_STATE_SOURCE_FASTTRACK)
+        if vm_artifacts_seq_no is not None:
+            sequence_number_file_path = os.path.join(conf.get_lib_dir(), SEQUENCE_NUMBER_FILE_NAME)
+            self._save_cache(sequence_number_file_path, ustr(vm_artifacts_seq_no))
+
+    def _set_fabric(self, incarnation=None, svd_seqNo = None):
+        path = os.path.join(conf.get_lib_dir(), GOAL_STATE_SOURCE_FILE_NAME)
+        self._save_cache(path, GOAL_STATE_SOURCE_FABRIC)
+        if incarnation is not None:
+            incarnation_file_path = os.path.join(conf.get_lib_dir(), INCARNATION_FILE_NAME)
+            self._save_cache(incarnation_file_path, ustr(incarnation))
+        if svd_seqNo is not None:
+            svd_seqNo_file_path = os.path.join(conf.get_lib_dir(), SVD_SEQNO_FILE_NAME)
+            self._save_cache(svd_seqNo_file_path, ustr(svd_seqNo))
+
+    def _remove_cache(self, file_name):
+        try:
+            path = os.path.join(conf.get_lib_dir(), file_name)
+            if os.path.exists(path):
+                os.remove(path)
+        except IOError as e:
+            fileutil.clean_ioerror(e, paths=path)
+            raise ProtocolError("Failed to remove cache: {0}".format(e))
+
+    def _set_cached_vm_id(self, cached_vm_id):
+        path = os.path.join(conf.get_lib_dir(), VM_ID_FILE_NAME)
+        self._save_cache(path, cached_vm_id)
+
+    def _save_cache(self, local_file, data):
```

Comment:
This is exactly the same as `WireClient.save_cache()`. Is there anyway we could avoid code duplication? Maybe move both to a Util and use that everywhere?

Reply:
I have spent effort to remove the duplication here, but the problem is that I have tests that some tests require the file to be written (ones that hit this class), while the previous tests are fine that MockWireClient doesn't implement this. I've tried creating a help class or calling the WireClient version, but both result in many test failures. Ultimately I decided that since this is a simple 5 line method, the effort isn't worth it.

Reply:
Ahh the feared Test cases. I totally understand the pain, no worries this is fine :)